### ISSUE TITLE
`reportUnknownArgumentType` complains about `Unknown` generic argument where `Any` parameter type is expected

### ISSUE THREAD

**Issue by Nathan-Fenner** [microsoft/pyright#5242]
**Describe the bug**
The strict `reportUnknownArgumentType` flag complains when a `list[Unknown]` is passed to a function accepting a `list[Any]`, even though the specific value of the `Unknown` type clearly doesn't matter.

For comparison, pyright doesn't complain when passing `list[Unknown]` to a function accepting `Any`. In both cases, the particular value of the `Unknown` type argument doesn't affect the inference/correctness of the rest of the program, since everything is assignable to/from `Any`.

**To Reproduce**
Example with list
```python
from typing import Any

def unknown_thing() -> object:
    ... # Produce a value of `object` type, so pyright doesn't narrow below too much

def print_anything(param: Any):
    print(param)

def print_list(param_list: list[Any]):
    print(param_list)

thing: object = unknown_thing()
if isinstance(thing, list):
    # thing now has narrowed type: list[Unknown]

    print_anything(thing)  # ok
    print_list(thing)      # error: Argument type is partially unknown
```

Example with generic class type
```python
from typing import Any, Generic, TypeVar

T = TypeVar("T")
class Wrapper(Generic[T]):
    ...

def unknown_thing() -> object:
    ...

def print_anything(param: Any):
    print(param)

def print_wrapper(param_wrapper: Wrapper[Any]):
    print(param_wrapper)

thing: object = unknown_thing()
if isinstance(thing, Wrapper):
    # thing has inferred type: Wrapper[Unknown]
    print_anything(thing)  # ok
    print_wrapper(thing)   # error: Argument type is partially unknown
```

**Expected behavior**
Just like `print_anything` is fine to call with `list[Unknown]` because it expects an `Any`, it should be fine to call `print_list` with `list[Unknown]` because it expects a `list[Any]`.

**VS Code extension or command-line**
I've confirmed both in VS Code and in CLI:
- VS Code pylance (`v2023.5.50`) using pyright `1.1.310`
- CLI `pyright 1.1.311`

**Additional context**
The `reportUnknownArgumentType` flag is helpful to me (so I don't want to just turn it off), because there's a lot of cases where we downcast using e.g. `isinstance` from some type and generic parameters become `Unknown` and we want to avoid assuming anything about them.

But there's also a lot of cases where after performing the downcast, we only want to pass to function that don't care about the generic argument type, so only ask for `Any`.

There are probably some more-complex cases involving subtyping/protocol/unions etc., but I'd be more than happy with a simple heuristic that only silences the error in the "exact match" case (i.e. where there's no subtyping/protocol/union ambiguities to worry about).

In my "real" use-case, usually the `isinstance` isn't narrowing from `object` to `list`, but more commonly `MyBaseClass` to `MyChildClass` where `MyChildClass` is generic but `MyBaseClass` is not:

```python
class MyBaseClass:
  ...
class MyChildClass(Generic[T]):
  ...

def get_base() -> MyBaseClass:
  ...
def process_as_child(child: MyChildClass[Any]):
  ...

base = get_base()
if isinstance(base, MyChildClass):
  process_as_child(base)
```

... but since the error is reported entirely after narrowing, I don't think this is strictly relevant to the complaint here.

**Comment by erictraut** [comment]
The `reportUnknownArgumentType` is a strict-mode diagnostic check and is somewhat opinionated in its behavior. The current behavior is "as designed", so I don't consider this a bug.

I'm reluctant to add heuristics or special casing that silence this diagnostic in some cases but not others. The current behavior is straightforward to document and understand.

I'll think about this a bit more, but my current inclination is to close it. We could reconsider it in the future if there is sufficient user feedback (e.g. in the form of upvotes or similar requests). To the best of my recollection, this is the first time someone has requested a change to the current behavior of this diagnostic check.

**Comment by erictraut** [comment]
If you find that this diagnostic check gives you many false positives, you might want to disable it in your code base. Making a legacy code base type check in "strict" mode is difficult. It's much easier to do when writing new code from scratch. In my team, we mandate that all new code type check in "strict" mode but older modules leave some of the strict-mode rules disabled.

Your suggestion (treating `Unknown` as never assignable) would break pretty much everything (including overload matching and constraint solving) because `Unknown` (which is just another name for an implicit `Any`) is assumed to be both a bottom and top type in the Python type system. This assumption permeates everything.

Even if that were a solution, I wouldn't want to add a switch that changes the type evaluation behavior. We've done that a couple of times in the past, and I've regretted it every time. It makes for a massive test matrix and significantly complicates diagnosis of reported bugs. We try very hard to keep the type evaluation rules the same for all pyright users. Differences should only be in which diagnostics are reported or suppressed. This is why the existing `reportUnknownXXX` family of diagnostic checks work the way they do.

We also generally don't like adding new switches that cannot be enabled by default in "strict" mode because they are not very discoverable, and almost no one benefits from them if they can't be enabled by default. The option you're proposing would fall into that category, so it's unlikely we'd add something like that.

**Comment by Nathan-Fenner** [comment]
> If you find that this diagnostic check gives you many false positives, you might want to disable it in your code base. Making a legacy code base type check in "strict" mode is difficult. It's much easier to do when writing new code from scratch. In my team, we mandate that all new code type check in "strict" mode but older modules leave some of the strict-mode rules disabled.

The problem that I'm running into here is that this isn't an issue with having a legacy codebase - I have the same issues in a brand new, pristine, fully-annotated, strict-checked codebase, because basic Python primitives like `isinstance` still produce `Unknown` types when dealing with generic classes. 

```python

from typing import Any, Generic, TypeVar

class BaseClass:
    ...

T = TypeVar("T")

class ChildClass(Generic[T]):
    def __init__(self, name: str, value: T):
        self.name: str = name
        self.value: T = value

def print_name(child: ChildClass[Any]) -> None:
    # This function can be called on any `ChildClass`, since we don't access `.value`.
    print(child.name)

def print_str_value_twice(child: ChildClass[str]) -> None:
    # This function can only be called on `ChildClass[str]`, because we rely on `.value` being `str`.
    print(child.value * 2)

def produce_base_class() -> BaseClass:
    ...

my_item = produce_base_class()
if isinstance(my_item, ChildClass):
    print_name(my_item)            # I *don't* want a type error here
    print_str_value_twice(my_item) # I *want* a type error here
```

- If `reportUnknownArgumentType` is enabled, then both calls produce type errors, which is problematic because it flags `print_name` even though that call is perfectly valid. 

- But if `reportUnknownArgumentType` is disabled, then neither call produces a type error, which is also bad.

Extending the existing heuristic for `reportUnknownArgument` would be the simplest approach to fixing this problem, but there are likely plenty of other directions to address it from as well.

**Comment by erictraut** [comment]
In your code sample above, you have a comment `# I *want* a type error here`. How would you address this issue if it arose in your code? Wouldn't this same technique work for the `# I *don't* want a type error here` example?

**Comment by Nathan-Fenner** [comment]
I want a type error in the first case, because it indicates a bug; the `isinstance` check establishes that you have a `list`, but nothing about its contents. The if is to introduce a new (domain-specific) cast/guard, like:

```python
def is_str_child(val: object) -> TypeGuard[ChildClass[str]]:
  return isinstance(val, ChildClass) and val.kind == "str-holder"
```


Guarding by `if is_str_child(my_item)` (or an equivalent) should be necessary to call `print_str_value_twice`, because that function only accepts a `ChildClass[str]`. Since a `ChildClass[Unknown]` might or might not be a `ChildClass[str]`, a diagnostic error makes sense in this case, and is currently only available via `reportUnknownArgumentType`.

In the case of calling `print_name`, this is not necessary; a check for `isinstance(my_item, ChildClass)` is sufficient because a `ChildClass[Unknown]` is always a `ChildClass[Any]`, since all types are convertible to/from `Any`. As a result, the diagnostic raised there is a false positive.

---

I want a diagnostic to be raised when calling `print_str_value_twice` with a `ChildClass[Unknown]` since the code needs to be fixed. I don't want a diagnostic when calling `print_name` with a `ChildClass[Unknown]` since this call is always valid.

The current check can't distinguish between these cases because it only considers the parameter type in _one_ special case, which is a parameter of type `Any`, and not other nested cases like, say, `ChildClass[Any]` or `Sequence[object]` which could e.g. safely accept `ChildClass[Unknown]` or `list[Unknown]` respectively.

**Comment by erictraut** [comment]
OK, thanks. I understand what you are looking for. 

The current behavior is working as intended, so I don't consider this a bug. Making the change you're asking for would add complexity and inconsistency to the current check, so I'm reluctant to make such a change. Your alternate suggestion for a switch to toggle between different evaluation behaviors is something we don't do for the reasons I mentioned above.

You're the first person to ask for this change, at least to the best of my recollection. I've personally used pyright to transform an untyped to a strictly-typed large Python code base, and this issue hasn't been a pain point or obstacle for me. Perhaps it's specific to your code base or coding style. 

I'm going to close this for now. I'm willing to revisit it in the future if there's sufficient demand from other pyright users (e.g. in the form of upvotes on this issue or similar requests).
